Add -Zannotate-moves for profiler visibility of move/copy operations (codegen)

jsgf opened the pull request:
**Note:** this is an alternative implementation of https://github.com/rust-lang/rust/pull/147206; rather than being a MIR transform, it adds the annotations closer to codegen. It's functionally the same but the implementation is lower impact and it could be more correct.

---

This implements a new unstable compiler flag `-Zannotate-moves` that makes move and copy operations visible in profilers by creating synthetic debug information. This is achieved with zero runtime cost by manipulating debug info scopes to make moves/copies appear as calls to `compiler_move<T, SIZE>` and `compiler_copy<T, SIZE>` marker functions in profiling tools.

This allows developers to identify expensive move/copy operations in their code using standard profiling tools, without requiring specialized tooling or runtime instrumentation.

The implementation works at codegen time. When processing MIR operands (`Operand::Move` and `Operand::Copy`), the codegen creates an `OperandRef` with an optional `move_annotation` field containing an `Instance` of the appropriate profiling marker function. When storing the operand, `store_with_annotation()` wraps the store operation in a synthetic debug scope that makes it appear inlined from the marker.

Two marker functions (`compiler_move` and `compiler_copy`) are defined in `library/core/src/profiling.rs`. These are never actually called - they exist solely as debug info anchors.

Operations are only annotated if:
   - We're generating debug info and the feature is enabled.
   - Meets the size threshold (default: 65 bytes, configurable via `-Zannotate-moves=SIZE`), and is non-zero
   - Has a memory representation

This has a very small size impact on object file size. With the default limit it's well under 0.1%, and even with a very small limit of 8 bytes it's still ~1.5%. This could be enabled by default.

<!-- homu-ignore:start -->
<!--
If this PR is related to an unstable feature or an otherwise tracked effort,
please link to the relevant tracking issue here. If you don't know of a related
tracking issue or there are none, feel free to ignore this.

This PR will get automatically assigned to a reviewer. In case you would like
a specific user to review your work, you can assign it to them by using

    r? <reviewer name>
-->
<!-- homu-ignore:end -->

jsgf committed:
Add -Zannotate-moves for profiler visibility of move/copy operations

This implements a new unstable compiler flag `-Zannotate-moves` that makes
move and copy operations visible in profilers by creating synthetic debug
information. This is achieved with zero runtime cost by manipulating debug
info scopes to make moves/copies appear as calls to `compiler_move<T, SIZE>`
and `compiler_copy<T, SIZE>` marker functions in profiling tools.

This allows developers to identify expensive move/copy operations in their
code using standard profiling tools, without requiring specialized tooling
or runtime instrumentation.

The implementation works at codegen time. When processing MIR operands
(`Operand::Move` and `Operand::Copy`), the codegen creates an `OperandRef`
with an optional `move_annotation` field containing an `Instance` of the
appropriate profiling marker function. When storing the operand,
`store_with_annotation()` wraps the store operation in a synthetic debug
scope that makes it appear inlined from the marker.

Two marker functions (`compiler_move` and `compiler_copy`) are defined
in `library/core/src/profiling.rs`. These are never actually called -
they exist solely as debug info anchors.

Operations are only annotated if the type:
   - Meets the size threshold (default: 65 bytes, configurable via
     `-Zannotate-moves=SIZE`)
   - Has a non-scalar backend representation (scalars use registers,
     not memcpy)

This has a very small size impact on object file size. With the default
limit it's well under 0.1%, and even with a very small limit of 8 bytes
it's still ~1.5%. This could be enabled by default.

saethlin commented:
Does this support cranelift? The advantage to doing such things in MIR is that it is much easier to support cranelift.

saethlin commented:
Ah, I'll respond on the MCP Zulip thread.

madsmtm commented on library/core/src/lib.rs:
```diff
@@ -279,6 +279,8 @@ pub mod num;
 pub mod intrinsics;
 pub mod mem;
+#[unstable(feature = "profiling_marker_api", issue = "none")]
+pub mod profiling;
```
I think you have to add something to `std/src/lib.rs` to get the API re-exported in `std` as well - that usually documents better. Unless this intentionally isn't done?

madsmtm replied:
> If they only ever appear as `core` then making them also appear in `std` could just be confusing.

But this is already the state for all debug info, no? If you call a `core` method that happens to be re-exposed in `std`, you'd still see `core` in your debugger.

> I was unsure about making it an exposed public module but functionally empty. Should that be deferred until it has some actual callable contents?

Unsure too, in the end, this is probably libs domain (?), I'll ping one of them for guidance once the MCP finishes.

madsmtm commented on src/doc/unstable-book/src/compiler-flags/annotate-moves.md:
```diff
@@ -0,0 +1,79 @@
+The annotation only applies to:
+- Types larger than the specified size threshold
+- Non-immediate types (those that would generate `memcpy`)
+- Operations that actually move/copy data (not ZST types)
```
What happens if you specify a threshold like 2 then? I assume that there wouldn't be any profiling functions called with that size. Or would you still see copies/moves to/from memory? If not, maybe we should issue a warning if a threshold < 64 is specified?

madsmtm replied:
It would also help if you had a test that tested for these kinds of things - e.g. tested that SIMD types aren't annotated, or tested that a `u32` move isn't annotated even if the threshold is very low.

madsmtm commented on tests/codegen-llvm/annotate-moves/annotate-moves-disabled.rs:
```diff
@@ -0,0 +1,33 @@
+//
+//@ compile-flags: -Copt-level=0 -g
+
+// Test that move/copy operations are NOT annotated when the flag is disabled
```
Nit: Remove empty `//`

```suggestion
// Test that move/copy operations are NOT annotated when the flag is disabled
//@ compile-flags: -Copt-level=0 -g
```

madsmtm commented on tests/codegen-llvm/annotate-moves/annotate-moves-integration.rs:
```diff
@@ -0,0 +1,114 @@
+//@ compile-flags: -Z annotate-moves=50 -Copt-level=0 -g
```
I'd like to also see tests of various types that follow different configurations than just `struct { array }` - maybe a raw array copy, maybe a copy of other primitives like `u32`, `(u128, u128)` or pointers.

Also interested in how things like `fn foo(x: &) { let _copy = *; }` would look, it's not clear to me that those wouldn't work differently from a stack copy?

jsgf replied:
> Also interested in how things like fn foo(x: &) { let _copy = *; } would look, it's not clear to me that those wouldn't work differently from a stack copy?

I added this and it turns out like regular Copy and is annotated with `compiler_copy`.

madsmtm commented on tests/codegen-llvm/annotate-moves/call-arg-scope.rs:
```diff
@@ -0,0 +1,33 @@
+//
+// Note: On most ABIs, large structs are passed by pointer even when written as "by value",
+// so there may not be an actual memcpy operation to attach compiler_move to. This test
+// mainly verifies that IF debug info is emitted, the call itself uses the source location.
```
It'd be nice with a test that does something like:
```rust
//@ revisions: by_pointer by_value
// Some target with an ABI that passes by pointer
//@ [by_pointer] compile-flags: --target x86_64-unknown-linux-gnu
//@ [by_pointer] needs-llvm-components: x86_64
// Some target with an ABI that passes by value (dunny one on top of my head)
//@ [by_value] compile-flags: --target abc-vendor-os
//@ [by_value] needs-llvm-components: abc
```

That would give me more confidence in that we're actually testing the right thing.

madsmtm commented on compiler/rustc_codegen_ssa/src/mir/operand.rs:
```diff
@@ -285,6 +290,22 @@ impl<'a, 'tcx, V: CodegenObject> OperandRef<'tcx, V> {
+    /// Store this operand into a place, applying move/copy annotation if present.
+    ///
+    /// This is the preferred method for storing operands, as it automatically
+    /// applies profiler annotations for tracked move/copy operations.
```
Q: Are there places where we are still calling `self.val.store`? If so, why?

jsgf replied:
There are no other calls to `self.val.store()` here, but there are still calls to `x.val.store()` around the place. Let me see if they also need to be changed.

madsmtm reviewed (changes requested):
Thanks! I recall working on Ruffle years back, and having expensive copies being an issue, something like this would definitely have helped back then.

I did a surface-level review, haven't truly dug into the implementation details of it yet, I'll probably wait 'till the MCP finishes before doing so.

@rustbot blocked (on MCP)

I'm also not 100% confident in my knowledge of things here, it's likely that I'll pass it off to saethlin afterwards to do an extra review on top.

*[View changes since this review](https://triagebot.infra.rust-lang.org/gh-changes-since/rust-lang/rust/147803/28c4c7d7abced7b35c49f38149b9cb6ea27dd2a6..06baf138ca2133a56dd24f49d7f8f59f9ddfa189)*

jsgf commented on src/doc/unstable-book/src/compiler-flags/annotate-moves.md:
```diff
@@ -0,0 +1,79 @@
+Where `<value>` can be:
+- A boolean: `true`, `false`, `yes`, `no`, `on`, `off`
+- A number: size threshold in bytes (e.g., `128`)
+- Omitted: enables with default threshold (65 bytes)
```
Yeah `65` seems a bit awkward, but I did want to make it "include all types <size> and above"; otherwise if you're trying to track down a specific case, knowing a structure is a specific size, and then having to remember "oh no set the limit to size-1" just seems like an unnecessary pitfall.

So if the default were "64" we'd end up including 64 byte things but not 63 which seems a bit weird.

jsgf replied:
Well, you're generally interested in sizes *greater* than some threshold, so it would  be >= <limit>.

madsmtm replied:
Ah, I might've confused myself, that might already be what the `if ty_size < size_limit { skip }` in `compiler/rustc_codegen_ssa/src/mir/operand.rs` does.

Would still be nice with a comment on `MOVE_ANNOTATION_DEFAULT_LIMIT` that discusses why `65` is a good default.

jsgf commented on compiler/rustc_codegen_ssa/src/mir/operand.rs:
```diff
@@ -1030,11 +1065,77 @@ impl<'a, 'tcx, Bx: BuilderMethods<'a, 'tcx>> FunctionCx<'a, 'tcx, Bx> {
+
+        // Look up the DefId for compiler_move or compiler_copy (skip if it's missing for some
+        // reason).
+        let def_id = tcx.get_diagnostic_item(kind)?;
```
Yeah I was also wondering about this. I decided on diagnostic just because it has no user-visible language semantics, and it has no functional effect on codegen, just the metadata around it. If one really wanted to come up with a rationalization, debuginfo is kinda like a warning message which is emitted by the compiler but doesn't prevent a functional artifact from being generated. (I know, it's a stretch.)

Lang-item really suggests to me that it has some special semantics associated with it.

But I'm fine either way.

I haven't really looked into the mechanics of lang-items, but is it just a matter of a different attribute on the function, and a different lookup function here?

jsgf commented on compiler/rustc_codegen_ssa/src/mir/operand.rs:
```diff
@@ -1014,7 +1038,18 @@ impl<'a, 'tcx, Bx: BuilderMethods<'a, 'tcx>> FunctionCx<'a, 'tcx, Bx> {
+                // Check if we should annotate this move/copy for profiling
+                let move_annotation = self.move_copy_annotation_instance(bx, ty, layout, kind);
+
+                OperandRef { move_annotation, ..self.codegen_consume(bx, place.as_ref()) }
```
Good point. I rearranged it a bit.

jsgf commented:
Updated to address review comments:
- refactored enable logic to push possibly expensive things later
- a few missed stores possibly needing annotation
- added CLI test
- added some more test cases

rust-log-analyzer commented:
The job **`aarch64-gnu-llvm-20-2`** failed! Check out the build log: [(web)](https://github.com/rust-lang/rust/actions/runs/18627869909/job/53108419029) [(plain enhanced)](https://triage.rust-lang.org/gha-logs/rust-lang/rust/53108419029) [(plain)](https://github.com/rust-lang/rust/commit/632004b6279d9f2a7df1311d1691e5965289ffc1/checks/53108419029/logs)

<details><summary><i>Click to see the possible cause of the failure (guessed by this bot)</i></summary>

```plain
   Compiling rustc_interface v0.0.0 (/checkout/compiler/rustc_interface)
error[E0599]: no variant or associated item named `Enable` found for enum `AnnotateMoves` in the current scope
   --> compiler/rustc_interface/src/tests.rs:767:45
    |
767 |     tracked!(annotate_moves, AnnotateMoves::Enable(Some(1234)));
    |                                             ^^^^^^ variant or associated item not found in `AnnotateMoves`
    |
help: there is a variant with a similar name
    |
767 |     tracked!(annotate_moves, AnnotateMoves::Enabled(Some(1234)));
    |                                                   +

[RUSTC-TIMING] rustc_driver_impl test:true 2.153
   Compiling rustc_public v0.1.0-preview (/checkout/compiler/rustc_public)
For more information about this error, try `rustc --explain E0599`.
```

</details>

madsmtm reviewed (commented):
I like the expanded tests!

> Squashed all the commits because they seemed uncontroversial

That's fine, feel free to squash the last one too as you see fit.

*[View changes since this review](https://triagebot.infra.rust-lang.org/gh-changes-since/rust-lang/rust/147803/27050c0d15af664cf43ce4b0badec230e0bfcac5..19dddf65bd2e1c0f2c0308d7c950402b87d0629b)*

madsmtm reviewed (approved):
I think I'd be good with this now, thanks!

r? saethlin could you check it over, I'm not that familiar with MIR? Feel free to r=me if you approve of it too.

*[View changes since this review](https://triagebot.infra.rust-lang.org/gh-changes-since/rust-lang/rust/147803/27050c0d15af664cf43ce4b0badec230e0bfcac5..3f7e412dcf1f6af50cd50da24bd2b99d789f37d5)*

jsgf commented on tests/codegen-llvm/annotate-moves/size-limit.rs:
```diff
@@ -0,0 +1,112 @@
+impl Clone for Struct99 {
+    // CHECK-LABEL: <size_limit::Struct99 as core::clone::Clone>::clone
+    fn clone(&self) -> Self {
+        // Should NOT be annotated since 80 < 100
```
99

saethlin commented:
Correct. The merge suite runs the tests on many more platforms, and probably runs some groups of tests that aren't run on PR CI. But the extra platforms is the usual cause of this.

What happens when you try to run the script?

saethlin commented:
@bors try jobs=test-various

bors commented:
:hourglass: Testing commit 5f29f11a4de9b731fe7e74434d3e2fba73cfc1f4 with merge 2da523c66d76f4bb170c3ebab55d927018a593d1...
<!-- homu: {"type":"BuildStarted","head_sha":"5f29f11a4de9b731fe7e74434d3e2fba73cfc1f4","merge_sha":"2da523c66d76f4bb170c3ebab55d927018a593d1"} -->

bors commented:
:hourglass: Testing commit 5f29f11a4de9b731fe7e74434d3e2fba73cfc1f4 with merge 59dfd4bf797133c356b4372055905952c92afd78...
<!-- homu: {"type":"BuildStarted","head_sha":"5f29f11a4de9b731fe7e74434d3e2fba73cfc1f4","merge_sha":"59dfd4bf797133c356b4372055905952c92afd78"} -->

saethlin commented:
Nooooo
@bors treeclosed p=100

saethlin commented:
@bors treeclosed=100

saethlin commented:
@bors retry yield to the close

bors commented:
:hourglass: Testing commit 5f29f11a4de9b731fe7e74434d3e2fba73cfc1f4 with merge e4d6a9a10820e0ba78fdb6c78fe0a387c650d5a2...
<!-- homu: {"type":"BuildStarted","head_sha":"5f29f11a4de9b731fe7e74434d3e2fba73cfc1f4","merge_sha":"e4d6a9a10820e0ba78fdb6c78fe0a387c650d5a2"} -->

bors commented:
:hourglass: Testing commit 5f29f11a4de9b731fe7e74434d3e2fba73cfc1f4 with merge bbb6f68e2888eea989337d558b47372ecf110e08...
<!-- homu: {"type":"BuildStarted","head_sha":"5f29f11a4de9b731fe7e74434d3e2fba73cfc1f4","merge_sha":"bbb6f68e2888eea989337d558b47372ecf110e08"} -->

madsmtm commented:
Regression / improvements both look spurious.

jsgf commented:
Yeah the only cost by default is the extra field which I guess could increase heap size. But no perf cost with the option off.
